### PR TITLE
Overriden document methods on a prototype leve (close #1827)

### LINKED ISSUE #1827: Vue Onsen UI and Testcafe e2e's do not work properly
### Are you requesting a feature or reporting a bug?
bug

### What is the current behavior?
Components render differently in normal Browser and Testcafe Browser environment. You'll find more details below.

### What is the expected behavior?
They should render the same way.

### How would you reproduce the current behavior (if this is a bug)?
Here's a repository containing the issue:
https://github.com/Orlandster/vue-testcafe-issue

Follow these steps to reproduce it:
-  `git clone https://github.com/Orlandster/vue-testcafe-issue.git`
- `cd vue-testcafe-issue`
- `yarn install` or `npm install`
- `yarn dev`
- Once it finished bundling open the url in the browser. If the port is not in use it should be: http://localhost:8080/ 

The output should look similar to this: https://jmp.sh/fIdqOZ8 -> (grey background and the icon)
Everything seems to work perfect! You can also notice that by analyzing the markup with the dev tools. For example if you debug the `ons-page` element it will look like this: https://jmp.sh/cIoMm2u

You can see there are 2 children `page__background` and `page__content`. In addition the console is also clear.

So far so good.

- As a next step open a new terminal window and execute `yarn test:e2e`. This will trigger the `testcafe` e2e test to start. Currently there is just one test located in the `test/e2e` folder. Basically it's not even a real test, it simply waits in order to give you time to debug everything.

This is the testcafe CLI command we execute: `testcafe chrome tests/e2e --screenshots ./screenshots --screenshots-on-fails --selector-timeout 20000 --assertion-timeout 10000`

- A Browser window with the application should open -> You will notice the application looks different smth. like this: https://jmp.sh/dlFEtge
- Open the dev tools -> You will also notice  there are no console errors. But if you debug the `ons-page` element you will see it looks different than before. Here's a screenshot: https://jmp.sh/9fo0rCV

It seem like the component wasn't rendered correctly. There is no `page__background` or `page__content` element.

> Notice: I think it's an issue between onsen and testcafe, but in addition I also want to mention, that I've also seen some weird behavior between those two and `async` code like `promises`.

Any help would be really appreciated!
THX!

#### Provide the test code and the tested page URL (if applicable)

Tested page URL: https://github.com/Orlandster/vue-testcafe-issue

### Specify your

* operating system: Mac OS X 10.13.6
* testcafe version: 0.22.0
* node.js version: 8.12.0
* chrome: 69.0.3497

### PR DESCRIPTION
**Pull request by miherlosev**

### CODE REVIEW

**Comment by miherlosev** [comment]
Waiting for https://github.com/DevExpress/testcafe-hammerhead/pull/1837

**Comment by Farfurix** [comment]
@miherlosev https://github.com/DevExpress/testcafe-hammerhead/pull/1842 is merged, need to rebase

**Comment by Farfurix** [comment]
@miherlosev https://github.com/DevExpress/testcafe-hammerhead/pull/1837 (Fix failed Internet Explorer 11 tests (NetworkError occurs after some Windows 10 update, iframe without src case)) is merged, need to rebase.

File: src/client/sandbox/iframe.js [hunk 1]
```diff
@@ -42,19 +42,18 @@ export default class IframeSandbox extends SandboxBase {
...
     _ensureIframeNativeMethodsForIE (iframe) {
         const contentWindow       = nativeMethods.contentWindowGetter.call(iframe);
+        const contentDocument     = nativeMethods.contentDocumentGetter.call(iframe);
         const iframeNativeMethods = contentWindow[INTERNAL_PROPS.iframeNativeMethods];
 
         if (iframeNativeMethods) {
-            const contentDocument = nativeMethods.contentDocumentGetter.call(iframe);
-
-            iframeNativeMethods.restoreDocumentMeths(contentDocument, contentWindow);
+            iframeNativeMethods.restoreDocumentMeths(contentWindow, contentDocument);
```
Comment (LavrovArtem) [inline src/client/sandbox/iframe.js]: Maybe just use `contentWindow.document`

File: src/client/sandbox/iframe.js [hunk 2]
```diff
@@ -115,7 +114,7 @@ export default class IframeSandbox extends SandboxBase {
         const isFFIframeUninitialized = isFirefox && contentWindow.document.readyState === 'uninitialized';
 
         return !isFFIframeUninitialized && !!contentDocument.documentElement ||
-               isIE && contentWindow[INTERNAL_PROPS.documentWasCleaned];
+            isIE && contentWindow[INTERNAL_PROPS.documentWasCleaned];
```
Comment (LavrovArtem) [inline src/client/sandbox/iframe.js]: discard this change

File: src/client/sandbox/native-methods.js [hunk 3]
```diff
@@ -20,42 +20,47 @@ class NativeMethods {
...
-        this.querySelector        = doc.querySelector || docProto.querySelector;
-        this.querySelectorAll     = doc.querySelectorAll || docProto.querySelectorAll;
+        this.createDocumentFragment = docPrototype.createDocumentFragment;
+        this.createElement          = docPrototype.createElement;
+        this.createElementNS        = docPrototype.createElementNS;
+
+        this.documentOpenPropOwnerName  = NativeMethods._getDocumentPropOwnerName(docPrototype, 'open');
+        this.documentClosePropOwnerName = NativeMethods._getDocumentPropOwnerName(docPrototype, 'close');
+        this.documentWritePropOwnerName = NativeMethods._getDocumentPropOwnerName(docPrototype, 'write');
+        this.documentWriteLnPropOwnerName = NativeMethods._getDocumentPropOwnerName(docPrototype, 'writeln');
```
Comment (LavrovArtem) [inline src/client/sandbox/native-methods.js]: formatting

File: src/client/sandbox/native-methods.js [hunk 4]
```diff
@@ -20,42 +20,47 @@ class NativeMethods {
...
+
+        this.documentOpen           = win[this.documentOpenPropOwnerName].prototype.open;
+        this.documentClose          = win[this.documentClosePropOwnerName].prototype.close;
+        this.documentWrite          = win[this.documentWritePropOwnerName].prototype.write;
+        this.documentWriteLn        = win[this.documentWriteLnPropOwnerName].prototype.writeln;
```
Comment (LavrovArtem) [inline src/client/sandbox/native-methods.js]: formatting

File: src/client/sandbox/node/document/index.js [hunk 5]
```diff
@@ -123,101 +126,114 @@ export default class DocumentSandbox extends SandboxBase {
...
-            const result = nativeMethods.documentOpen.apply(document, args);
+                // NOTE: Chrome does not remove the "%hammerhead%" property from window
+                // after document.open call
+                const objectDefinePropertyFn = window[INTERNAL_PROPS.hammerhead]
+                    ? window[INTERNAL_PROPS.hammerhead].nativeMethods.objectDefineProperty
+                    : window.Object.defineProperty;
 
-            if (isIE)
-                nativeMethods.refreshIfNecessary(document, window);
+                objectDefinePropertyFn
```
Comment (LavrovArtem) [inline src/client/sandbox/node/document/index.js]: `objectDefinePropertyFn(window, ...`

**Pull request merged**